Git submodules remote tracking

neutronth opened the pull request:
Add Git submodules remote tracking

* Always update the registered submodules using the remote-tracking branch
  instead of the superproject's recorded SHA-1.
  Only apply to the submodules' name that listed in the
  `--submodules-remote-tracking` parameter, otherwise default behavior will be applied.

  - Use case:
    Update the latest registered configurations submodule from remote repository
  - git-sync parameter:
    ./git-sync ... --submodules-remote-tracking="deployment-configs,app-configs"

* Fixes #265

neutronth commented:
/check-cla

thockin commented on cmd/git-sync/main.go:
```diff
@@ -56,6 +57,9 @@ var flDepth = flag.Int("depth", envInt("GIT_SYNC_DEPTH", 0),
 	"Configure submodule behavior - options are 'recursive', 'shallow' and 'off'.")
+var flSubmodulesRemoteTracking = flag.String("submodules-remote-tracking",
+	envString("GIT_SYNC_SUBMODULES_REMOTE_TRACKING", ""),
+	"Configure submodules list for remote tracking.")
```
This needs to explain the syntax and meaning a little

neutronth replied:
Try my best to explain in this commit https://github.com/kubernetes/git-sync/pull/266/commits/c508129251286b7b623e71f84e6bc73a885afa6c

Please advise.

thockin commented on cmd/git-sync/main.go:
```diff
@@ -585,6 +580,207 @@ func addWorktreeAndSwap(ctx context.Context, gitRoot, dest, branch, rev string,
+	}
+
+	// Update submodules with remote tracking enabled
+	if *flSubmodulesRemoteTracking != "" {
```
please pass this in as an arg - I know it is a bit tedious, but only main() should use raw flags (and yes I know that is not always true already :)

neutronth replied:
Propose the `submoduleOptions type` which encapsulates the options and passing around with this variable instead of the direct access to the `flSubmodules*` which should be only handled in the `main()` as your suggestion.

Proposed patch in this commit https://github.com/kubernetes/git-sync/pull/266/commits/dcf07c19567111761108bc64b8c7c789c3627bdd

thockin reviewed (commented):
Hi!  Thanks for this PR.

There's a lot going on in here that I don't quite understand, and I will have to patch this in and play with it to understand better.  I may have feedback when I get to that.  Any additional context you can specify here (comments or just in the PR or commit descriptions) would help a ton.  If I am being honest, it may be a few weeks before I get a chance to do this work.

I have a bigger problem, though.  I've been working on a v4 branch (not published yet!) which does a MASSIVE overhaul of the internals, cleans up flags, logs, etc.  This includes some breaking changes.

This patch is complex enough that I one of two things happens:
1) I merge this to master and manually reimplement it in the new structure
2) I ask you to make this work in v4 (either only in v4 or in both)

How do you feel about that?

neutronth commented:
Thank you so much for your review.

> This patch is complex enough that I one of two things happens:
> 
>     1. I merge this to master and manually reimplement it in the new structure

I have re-based the branch with current master to keep in sync with the current code-base and propose another
refactor codes, sorry about that if it makes more complexity.
however, I think, the long arguments passing seems to be a bit of mess.

>     2. I ask you to make this work in v4 (either only in v4 or in both)

I'll checkout the `v4` branch and try to apply my patches on top of it. It may takes times but I'll update the status periodically.

> How do you feel about that?

It's very OK, I'll handle the `2.` and keep in touch here.

thockin commented:
The branch to track is https://github.com/thockin/git-sync/tree/use-git-mirror

thockin commented:
But I should add that I have been running open-loop. so some of it will need to be adjusted after review.

neutronth commented:
> The branch to track is https://github.com/thockin/git-sync/tree/use-git-mirror

Work in progress is https://github.com/neutronth/git-sync/tree/based-thockin-use-git-mirror/git-submodules-remote-tracking

jonathonbattista commented:
Any update on this?

jonathonbattista commented:
@neutronth You are a god-send!

Just tested your PR and it is working as expected!

Thanks!

thockin commented:
Hi, I apologize - it's still in my queue

On Wed, Dec 2, 2020 at 3:57 PM Kubernetes Prow Robot <
notifications@github.com> wrote:

> @neutronth <https://github.com/neutronth>: PR needs rebase.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/git-sync/pull/266#issuecomment-737567801>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKWAVCMTSE6GPNOB4S6PDDSS3IABANCNFSM4QRP4E6Q>
> .
>

neutronth commented:
Hi, sorry that I reply late.

The proposed work in progress [1] that based [2] has a major change from the initial patch in this PR.

I clone the concept from the `webhook` invocation that splits the go-routine task and separates the code base into the new file from main. Therefore, the main file has been touched in only a few points of code hook.
Just introduced a new file which included all the logic that handle the Git submodules remote tracking synchronization.
Hopefully, it may help the reviewers to walk through the code more easily.

For the status,
I just saw the new `release-3.x` default branch, please give me some times to prepare the patch based on this branch.

Thanks for your all for feedbacks.

---
[1] https://github.com/neutronth/git-sync/tree/based-thockin-use-git-mirror/git-submodules-remote-tracking
[2] https://github.com/thockin/git-sync/tree/use-git-mirror

jonathonbattista commented:
@neutronth Can you please add logging for checking out the submodule?

With this code it doesn't say which commit is being used per submodule 


```
git-sync-clone INFO: detected pid 1, running init handler
git-sync-clone I1211 19:07:13.813870      11 main.go:459]  "level"=0 "msg"="starting up"  "args"=["/git-sync"] "pid"=11
git-sync-clone I1211 19:07:13.814052      11 main.go:905]  "level"=0 "msg"="cloning repo"  "origin"="git@github.com:XXXX" "path"="/git"
git-sync-clone I1211 19:07:14.453336      11 main.go:615]  "level"=0 "msg"="syncing git"  "hash"="xxxxxxxx" "rev"="HEAD"
git-sync-clone I1211 19:07:15.044222      11 main.go:636]  "level"=0 "msg"="adding worktree"  "branch"="origin/xxxx" "path"="/git/rev-xxxxxxx"
git-sync-clone I1211 19:07:15.049459      11 main.go:659]  "level"=0 "msg"="reset worktree to hash"  "hash"="xxxxxxx" "path"="/git/rev-xxxxxx"
git-sync-clone I1211 19:07:15.049498      11 main.go:700]  "level"=0 "msg"="updating submodules"  
```

jonathonbattista commented:
Looks like it is not properly tracking the submodule branch as we originally suspected.

It dies after at **submoduleNameFromPath** after **updateSubmodulesWithRemoteTracking**
```
git-sync-clone  11 main.go:1077]  "level"=0 "msg"="running command"  "cmd"="git submodule --quiet foreach pwd" "cwd"="/git/rev-84fbd4d87f3522e6bccfcd5217b00316d0a36c60"
git-sync-clone  11 main.go:1098]  "level"=0 "msg"="command result"  "stderr"="" "stdout"="/git/rev-84fbd4d87f3522e6bccfcd5217b00316d0a36c60/plugins\n"
git-sync-clone  11 main.go:1077]  "level"=0 "msg"="running command"  "cmd"="git config --file .gitmodules --get-regexp .path$" "cwd"="/git/rev-84fbd4d87f3522e6bccfcd5217b00316d0a36c60"
git-sync-clone  11 main.go:1098]  "level"=0 "msg"="command result"  "stderr"="" "stdout"="submodule.plugins.path plugins\n"
git-sync-clone  11 main.go:843]  "level"=0 "msg"="looking up submodule name"  "configPath"="submodule.plugins.path plugins" "submodulePath"="plugins"
git-sync-clone  11 main.go:851]  "level"=0 "msg"="found submodule"  "submoduleName"="plugins"
git-sync-clone  11 main.go:1077]  "level"=0 "msg"="running command"  "cmd"="ln -snf rev-84fbd4d87f3522e6bccfcd5217b00316d0a36c60 tmp-link" "cwd"="/git"
git-sync-clone  11 main.go:1098]  "level"=0 "msg"="command result"  "stderr"="" "stdout"=""
git-sync-clone  11 main.go:1077]  "level"=0 "msg"="running command"  "cmd"="mv -T tmp-link Airflow" "cwd"="/git"
git-sync-clone  11 main.go:1098]  "level"=0 "msg"="command result"  "stderr"="" "stdout"=""
```

neutronth commented:
@jonathonbattista

> Looks like it is not properly tracking the submodule branch as we originally suspected.
>
> While **git submodule update --init** should be the one-liner to init and update the submodule to track the remote branch, it hardly ever succeeds on anything other than the default branch.
> 
> You need to run **git submodule init** first, then **git submodule update**, then **git submodule update --remote --recursive --depth 5 xxx** . Otherwise, you can run into errors like this:
> 
> ```
> "msg"="too many failures, aborting" "error"="Run(git submodule update --remote --recursive --depth 5 xxx): exit status 1: { stdout: \"\", stderr: \"fatal: Needed a single revision\\nUnable to find current origin/xxx revision in submodule path 'xxx'\\n\" }"  "failCount"=0
> ```

I could reproduce this also for the **shallow clone** with **--depth=1** for git-sync in the **test-e2e.sh**,
and thanks for your initial guideline to get rid of this issue.

```
git submodule init
git submodule update
git submodule update --init --recursive --depth=1
```
It works.

> I have tested adding these changes and it properly/consistently sets up a submodule to track a remote branch (particularly a branch other than the default branch).

Agree.
however, IMO, it would be better to introduce the new PR for this case, I think this issue would be occurred with the
project that using the **submodules with branch** initialization without the remote tracking as well.
I didn't see the test case that cover this case yet.
I might be wrong, I'll try to reproduce it with the **e2e-test** to verify that.

FYI, I've ported the code from [1] for the `master, release-3.x` and updated in this PR. The **test-e2e.sh** looks fine
without the shallow clone. Therefore, you could apply your patch on top to fix it.
For the previous code, I moved it to [2]

Thanks again for your feedback

---
[1] https://github.com/neutronth/git-sync/tree/based-thockin-use-git-mirror/git-submodules-remote-tracking
[2] https://github.com/neutronth/git-sync/tree/git-submodules-remote-tracking-v0

thockin commented:
I have not forgotten this PR!

nirutgupta commented:
Let's test and merge this.

thockin commented:
This PR won't fly as-is.  Discussion in #265
